File: C++/showenv.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <iomanip>
#include <cctype>
#include <sys/ioctl.h>
#include <unistd.h>

extern char **environ;

struct EnvVar {
    std::string key;
    std::string value;
};

// Case-insensitive substring finder helper
bool contains_keyword(std::string data, std::string to_find) {
    std::transform(data.begin(), data.end(), data.begin(), [](unsigned char c) { return std::tolower(c); });
    std::transform(to_find.begin(), to_find.end(), to_find.begin(), [](unsigned char c) { return std::tolower(c); });
    return data.find(to_find) != std::string::npos;
}

bool is_ip(const std::string &s) {
    int octets = 0, digits = 0;
    for (char c : s) {
        if (std::isdigit(c)) {
            if (++digits > 3) return false;
        } else if (c == '.') {
            if (digits == 0 || ++octets > 3) return false;
            digits = 0;
        } else {
            return false;
        }
    }
    return (octets == 3 && digits > 0);
}

void colorize_values(const std::string &val) {
    if (contains_keyword(val, "true") && val.length() == 4) { std::cout << "\e[32m" << val << "\e[0m"; return; }
    if (contains_keyword(val, "stderr") && val.length() == 6) { std::cout << "\e[31m" << val << "\e[0m"; return; }
    if (is_ip(val)) { std::cout << "\e[92m" << val << "\e[0m"; return; }

    if (contains_keyword(val, "truecolor")) {
        std::cout << "\e[31mt\e[32mr\e[33mu\e[36me\e[94mc\e[35mo\e[92ml\e[94mo\e[36mr\e[0m";
    } else if (contains_keyword(val, "256color")) {
        std::cout << "\e[31m2\e[32m5\e[33m6\e[36mc\e[94mo\e[35ml\e[92ml\e[94mo\e[36mr\e[0m";
    } else if (contains_keyword(val, "ubuntu")) {
        std::cout << "\e[38;5;202m" << val << "\e[0m";
    } else if (contains_keyword(val, "redhat"))  { std::cout << "\e[91m" << val << "\e[0m"; }
      else if (contains_keyword(val, "fedora"))  { std::cout << "\e[96m" << val << "\e[0m"; }
      else if (contains_keyword(val, "mint"))    { std::cout << "\e[92m" << val << "\e[0m"; }
      else if (contains_keyword(val, "zorin"))   { std::cout << "\e[97m" << val << "\e[0m"; }
      else if (contains_keyword(val, "wayland")) { std::cout << "\e[93m" << val << "\e[0m"; }
      else { std::cout << val; }
}

int main() {
    std::vector<EnvVar> env_list;
    size_t max_len = 0;

    for (char **env = environ; *env != nullptr; env++) {
        std::string raw(*env);
        size_t eq_pos = raw.find('=');
        if (eq_pos != std::string::npos) {
            std::string key = raw.substr(0, eq_pos);
            std::string value = raw.substr(eq_pos + 1);
            max_len = std::max(max_len, key.length());
            env_list.push_back({key, value});
        }
    }

    std::sort(env_list.begin(), env_list.end(), [](const EnvVar &a, const EnvVar &b) {
        return a.key < b.key;
    });

    std::cout << "\n\e[2K\e[93;44m  Environment Variables\e[0m\n";
    for (const auto &ev : env_list) {
        std::cout << "\e[1m\e[33;41m " << std::setw(max_len) << ev.key << " \e[0m = ";
        colorize_values(ev.value);
        std::cout << "\n";
    }
    std::cout << "\e[44m\e[2K\e[0m\n";
    return 0;
}

File: Rust/src/bin/showenv.rs
use std::env;
use std::io::{self, Write};

fn contains_keyword(data: &str, keyword: &str) -> bool {
    data.to_lowercase().contains(&keyword.to_lowercase())
}

fn is_ip(s: &str) -> bool {
    let mut octets = 0;
    let mut digits = 0;
    for c in s.chars() {
        if c.is_ascii_digit() {
            digits += 1;
            if digits > 3 { return false; }
        } else if c == '.' {
            if digits == 0 { return false; }
            digits = 0;
            octets += 1;
            if octets > 3 { return false; }
        } else {
            return false;
        }
    }
    octets == 3 && digits > 0
}

fn colorize_values(val: &str) {
    if contains_keyword(val, "true") && val.len() == 4 { print!("\x1b[32m{}\x1b[0m", val); return; }
    if contains_keyword(val, "stderr") && val.len() == 6 { print!("\x1b[31m{}\x1b[0m", val); return; }
    if is_ip(val) { print!("\x1b[92m{}\x1b[0m", val); return; }

    if contains_keyword(val, "truecolor") {
        print!("\x1b[31mt\x1b[32mr\x1b[33mu\x1b[36me\x1b[94mc\x1b[35mo\x1b[92ml\x1b[94mo\x1b[36mr\x1b[0m");
    } else if contains_keyword(val, "256color") {
        print!("\x1b[31m2\x1b[32m5\x1b[33m6\x1b[36mc\x1b[94mo\x1b[35ml\x1b[92ml\x1b[94mo\x1b[36mr\x1b[0m");
    } else if contains_keyword(val, "ubuntu") {
        print!("\x1b[38;5;202m{}\x1b[0m", val);
    } else if contains_keyword(val, "redhat")  { print!("\x1b[91m{}\x1b[0m", val); }
      else if contains_keyword(val, "fedora")  { print!("\x1b[96m{}\x1b[0m", val); }
      else if contains_keyword(val, "mint")    { print!("\x1b[92m{}\x1b[0m", val); }
      else if contains_keyword(val, "zorin")   { print!("\x1b[97m{}\x1b[0m", val); }
      else if contains_keyword(val, "wayland") { print!("\x1b[93m{}\x1b[0m", val); }
      else { print!("{}", val); }
}

fn main() {
    let mut env_list: Vec<(String, String)> = env::vars().collect();
    
    // Sort environment variables alphabetically by key
    env_list.sort_by(|a, b| a.0.cmp(&b.0));

    let max_len = env_list.iter().map(|(k, _)| k.len()).max().unwrap_or(0);

    println!("\n\x1b[2K\x1b[93;44m  Environment Variables\x1b[0m");
    for (key, value) in env_list {
        print!("\x1b[1m\x1b[33;41m {:>width$} \x1b[0m = ", key, width = max_len);
        colorize_values(&value);
        println!();
        let _ = io::stdout().flush(); // Ensure line renders immediately
    }
    println!("\x1b[44m\x1b[2K\x1b[0m\n");
}

